### PR TITLE
Make core config accessible in dict get way 

### PR DESCRIPTION
**Pull request by MerelTheisenQB**
## Description
https://github.com/kedro-org/kedro/issues/1863

## Development notes

- Changed `AbstractConfigLoader` to inherit from `UserDict` 
- Moved regex patterns for core config locations from `KedroContext` to `ConfigLoader` and `TemplatedConfigLoader`
- Core config patterns are stored in a dictionary so they can be accessed with dict syntax
- `config_patterns` is added as an argument to the `ConfigLoader` and `TemplatedConfigLoader`

## To discuss
Is allowing users to access config in this way something we want for the future `0.19.0` version as well? How does this allow for using non file based config, e.g stored in a database?

**Discussion summary**

- This change opens up to ability for users to change the config file patterns.
- The pattern paths are added to the `ConfigLoader` and `TemplatedConfigLoader` and not to the `AbstractConfigLoader` to not make add assumptions about file based config in the base abstract class.
- This also means that (in the future) users will be able to point to a database. The implementation of reading from databases would not happen on the Kedro side: it's not possible for us to build functionality for all possible databases that users have. However, the improvements to `ConfigLoader`will enable users to implement their own DB access so they can load their config. 
- In `0.19.0` we will remove the custom `.get()` method and rely on `OmegaConf` for loading the actual config.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [x] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [x] Added tests to cover my changes

<a href="https://gitpod.io/#https://github.com/kedro-org/kedro/pull/1870"></a>

### CODE REVIEW

File: kedro/config/abstract_config.py [hunk 1]
```diff
@@ -1,11 +1,11 @@
 """This module provides ``kedro.abstract_config`` with the baseline
 class model for a `ConfigLoader` implementation.
 """
-from abc import ABC, abstractmethod
+from abc import abstractmethod
 from typing import Any, Dict
 
 
-class AbstractConfigLoader(ABC):
+class AbstractConfigLoader(dict):
```
Comment (noklam) [inline kedro/config/abstract_config.py:8]: I think it would be better to inherit from `UserDict` instead of the `dict`, there are some traps with inheriting from the Python standard `dict` class.

It would also be more consistent, see `_ProjectLogging` that we are already using `UserDict`.
https://github.com/kedro-org/kedro/blob/4db0ae3be611f60f3c98da2501364d5549c784f2/kedro/framework/project/__init__.py#L199-L208

Fluent Python has a detail explanation for it.

See these two relevant chapter
* https://learning.oreilly.com/library/view/fluent-python-2nd/9781492056348/ch14.html#subclass_builtin_woes
* https://learning.oreilly.com/library/view/fluent-python-2nd/9781492056348/ch03.html#missing_method

File: kedro/config/config.py [hunk 2]
```diff
@@ -86,18 +86,32 @@ def __init__(
...
         self.base_env = base_env
         self.default_run_env = default_run_env
 
+        core_config_patterns = {
+            "catalog": ["catalog*", "catalog*/**", "**/catalog*"],
+            "parameters": ["parameters*", "parameters*/**", "**/parameters*"],
+            "credentials": ["credentials*", "credentials*/**", "**/credentials*"],
+            "logging": ["logging*", "logging*/**", "**/logging*"],
+        }
+        self.config_patterns = core_config_patterns
```
Comment (noklam) [inline kedro/config/config.py:98]: ⭐️  I like this

**Review by noklam — COMMENTED** [review]
I like the direction of this change and it should make it easier towards customisable path for all configurations ⭐️. 

This is just something always in my mind since this question almost come up every week and I always has to point people to make a copy & paste `CustomTemplateConfigLoader`. Would be great if we can do something about it while we are making changes to these classes. This is definitely out of scope of this PR, but I just want bring it out so we can see if there are any opportunity to improve it.
https://github.com/kedro-org/kedro/issues/1527

**Review by AntonyMilneQB — COMMENTED** [review]
tl;dr: I think this is a good start, definitely an improvement or where things are now and probably fine to merge after thinking about the inheritance a bit more, but I'm not sure exactly how the next step will work. I'd feel more confident about it if @idanov could weight in with his thoughts!

I think there were a few major requirements here:
* enable access through getitem syntax:`config_loader["params"]` etc. ✅ 
* change all our `.get` calls to `__getitem__` ✅ 
* maintain backwards compatibility so that 3rd party `.get` calls don't need to change ✅ 

So I think that in basic terms we've achieved what we wanted to here.

But I'm left with several questions about exactly how it should be done:
1. what should `AbstractConfigLoader` inherit from? IMO at the moment it should just be `ABC` still. I would personally leave the dictionary inheritance for the future if and when it becomes a purely dict-like object
2. where should the `__getitem__` definition go - abstract config loader or the implementations? At the moment it's just copied and pasted between the implementations which seems a bit unDRY.
3. likewise do `core_config_patterns` belong in the abstract class or the implementations?
4. is the intention to remove `.get` entirely in the future?

tbh these points are all kind of moot though given that it's not at all clear what the abstract config loader is actually for currently...

More practically, what is actually the expected user journey following this change if I want to (a) add spark config and (b) add `*parameters` as a glob pattern? This is  https://github.com/kedro-org/kedro/issues/1864 but it feels a bit hard to judge the proposed change here without understanding how that would work. Currently I think it would be something like this:
```
class MyConfigLoader(ConfigLoader):
    def __init__(...):
         super().__init__(...)
         self.core_config_patterns.append("*parameters")
         self.core_config_patterns["spark"] = ["spark*", ...]
         # really we should have another object that's not called "core" to add spark to
```

I believe @idanov's plan was instead of doing class inheritance to make this patterns dictionary directly accessible in settings.py. But I don't quite see how that works. Would we then want to pass the patterns into the constructor of `ConfigLoader`? And how would that eventually enable different sort of config, e.g. database vs. file? We don't need to have it all worked out now, but I'd personally feel more comfortable with this change if I could see better what the ultimate consequence will look like.

File: kedro/config/abstract_config.py (hunk 1 above)
Comment (MerelTheisenQB) [reply kedro/config/abstract_config.py:8]: I don't remember this 100% anymore, but I think initially the plan was to have [`AbstractConfigLoader` inherit from `MutableMapping`](https://treyhunner.com/2019/04/why-you-shouldnt-inherit-from-list-and-dict-in-python/#Abstract_base_classes:_they%E2%80%99ll_help_you_quack_like_a_duck) but then it ended up being a [lot of overwriting of stuff we don't need.](https://github.com/kedro-org/kedro/commit/6dc6eb151bc9112752d6fe669912c689a1921732) And then the decision was made to inherit from `dict` directly. This was all over Slack and not very coherent. 

Thinking about this again now, I feel like you might be right and we don't actually need to inherit from `dict` at all since we're overwriting `get` and `__getitem__` anyway.

File: kedro/framework/context/context.py [hunk 3]
```diff
@@ -240,10 +240,7 @@ def params(self) -> Dict[str, Any]:
                 extra parameters passed at initialization.
         """
         try:
-            # '**/parameters*' reads modular pipeline configs
-            params = self.config_loader.get(
-                "parameters*", "parameters*/**", "**/parameters*"
-            )
+            params = self.config_loader["parameters"]
```
[the remark this answers is not in the document]
Comment (MerelTheisenQB) [reply kedro/framework/context/context.py:243]: Yes exactly! This is introduces as a non-breaking change and so we can't completely change/remove `get` yet.

**Comment by MerelTheisenQB** [comment]
As always, thank you for your thorough comments and feedback @AntonyMilneQB ⭐ 

> But I'm left with several questions about exactly how it should be done:
> 
> 1. what should `AbstractConfigLoader` inherit from? IMO at the moment it should just be `ABC` still. I would personally leave the dictionary inheritance for the future if and when it becomes a purely dict-like object

Yes I think that would make sense, see my earlier comment: https://github.com/kedro-org/kedro/pull/1870#discussion_r982258144 


> 2. where should the `__getitem__` definition go - abstract config loader or the implementations? At the moment it's just copied and pasted between the implementations which seems a bit unDRY.
> 3. likewise do `core_config_patterns` belong in the abstract class or the implementations?

These are good questions, my feeling is it should be in the loader implementations because it's too specific to be in the abstract class. As in the mapping from "params" to the parameter regexes isn't something the abstract implementation should care about or anyone inheriting from it. But I also agree that it's not clear what the role of the `AbstractConfigLoader` is anymore.. 

> 4. is the intention to remove `.get` entirely in the future?

Yes as far as I understand it, and instead just leverage the `dict` get methods. 

> More practically, what is actually the expected user journey following this change if I want to (a) add spark config and (b) add `*parameters` as a glob pattern? This is #1864 but it feels a bit hard to judge the proposed change here without understanding how that would work. Currently I think it would be something like this:
> 
> ```
> class MyConfigLoader(ConfigLoader):
>     def __init__(...):
>          super().__init__(...)
>          self.core_config_patterns.append("*parameters")
>          self.core_config_patterns["spark"] = ["spark*", ...]
>          # really we should have another object that's not called "core" to add spark to
> ```
> 
> I believe @idanov's plan was instead of doing class inheritance to make this patterns dictionary directly accessible in settings.py. But I don't quite see how that works. Would we then want to pass the patterns into the constructor of `ConfigLoader`? And how would that eventually enable different sort of config, e.g. database vs. file? We don't need to have it all worked out now, but I'd personally feel more comfortable with this change if I could see better what the ultimate consequence will look like.

I coded this up in an older (less clean) branch: https://github.com/kedro-org/kedro/compare/abstract-config-loader, but essentially you would have something like this in `settings.py`:

```
CONFIG_LOADER_CLASS = ConfigLoader
CONFIG_LOADER_ARGS = {
   "custom_patterns":{"spark" : ["spark*/"]}
}
```

and inside the `ConfigLoader`
```
 if custom_patterns:
    core_patterns.update(custom_patterns)
 self.config_patterns = core_patterns
```
So you can add whatever custom config you have as well as overwrite any of the "core" config locations.

**Comment by noklam** [comment]
> 1. what should AbstractConfigLoader inherit from? IMO at the moment it should just be ABC still. I would personally leave the dictionary inheritance for the future if and when it becomes a purely dict-like object

I just want to point out that with this PR, we already assume `config_loader` is a dict-like object. If we are going with `ABC`, then we should also add `__getitem__` as the class method.

@MerelTheisenQB @AntonyMilneQB 
Example:
```
# context.py
 params = self.config_loader["parameters"]
```

File: docs/source/kedro_project_setup/configuration.md [hunk 4]
```diff
@@ -334,7 +334,7 @@ conf_path = str(project_path / settings.CONF_SOURCE)
 conf_loader = ConfigLoader(conf_source=conf_path, env="local")
 
 try:
-    credentials = conf_loader.get("credentials*", "credentials*/**")
+    credentials = conf_loader["credentials*"]
```
Comment (idanov) [inline docs/source/kedro_project_setup/configuration.md:337]:
```suggestion
    credentials = conf_loader["credentials"]
```

File: kedro/config/abstract_config.py [hunk 5]
```diff
@@ -19,12 +19,13 @@ def __init__(
...
         **kwargs  # pylint: disable=unused-argument
     ):
+        super().__init__()
         self.conf_source = conf_source
         self.env = env
         self.runtime_params = runtime_params
 
     @abstractmethod  # pragma: no cover
-    def get(self) -> Dict[str, Any]:
+    def get(self) -> Dict[str, Any]:  # type: ignore
```
Comment (idanov) [inline kedro/config/abstract_config.py:28]: Is this still needed if we are inheriting from `dict` or `UserDict`? I suppose we can completely remove it from here, right?

File: kedro/framework/context/context.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (idanov) [reply kedro/framework/context/context.py:243]: I wonder whether it's worth opening a follow up PR straight to develop with the breaking change so we don't forget to do this when we release 0.19?

File: kedro/config/config.py (hunk 2 above)
Comment (idanov) [inline kedro/config/config.py:98]: Can we include a `config_patterns` argument in the `__init__`, so those patterns are configurable through `settings.py`? Then the code here could look like:

```
self.config_patterns = {
    "catalog": ["catalog*", "catalog*/**", "**/catalog*"],
    "parameters": ["parameters*", "parameters*/**", "**/parameters*"],
    "credentials": ["credentials*", "credentials*/**", "**/credentials*"],
    "logging": ["logging*", "logging*/**", "**/logging*"],
}

self.config_patterns.update(config_patterns or {})
```

File: kedro/config/templated_config.py [hunk 6]
```diff
@@ -114,6 +114,14 @@ def __init__(
                 obtained from the globals_pattern. In case of duplicate keys, the
                 ``globals_dict`` keys take precedence.
         """
+        core_config_patterns = {
+            "catalog": ["catalog*", "catalog*/**", "**/catalog*"],
+            "parameters": ["parameters*", "parameters*/**", "**/parameters*"],
+            "credentials": ["credentials*", "credentials*/**", "**/credentials*"],
+            "logging": ["logging*", "logging*/**", "**/logging*"],
+        }
+        self.config_patterns = core_config_patterns
```
Comment (idanov) [inline kedro/config/templated_config.py:123]: Same here, see the comment above for `ConfigLoader`.

**Review by idanov — COMMENTED** [review]
Awesome, such a small code change enabling so much extra functionality!

Could you add a very short description of this to the RELEASE.md?

File: kedro/framework/context/context.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (MerelTheisenQB) [reply kedro/framework/context/context.py:243]: We decided to wait with changes for `0.19.0` for other issues to make sure the `main` and `develop` branch don't diverge so much as with `0.18.0`. I'll make sure to add GH issues for any follow up actions that come out of this work.

File: kedro/config/abstract_config.py (hunk 5 above)
Comment (MerelTheisenQB) [reply kedro/config/abstract_config.py:28]: I think the bigger question raised by @noklam and @AntonyMilneQB was if we should actually be inheriting from `dict` or `UserDict` or just keep it to `ABC`: https://github.com/kedro-org/kedro/pull/1870#discussion_r981264452 
https://github.com/kedro-org/kedro/pull/1870#issuecomment-1261131796

I don't have a strong opinion on this, but feel that the intention to make the `ConfigLoader` classes more dict like is clearer if we inherit from `dict`/`UserDict`.

What do you think @idanov ?

File: kedro/config/config.py (hunk 2 above)
Comment (MerelTheisenQB) [reply kedro/config/config.py:98]: Yes can do. Originally I intended to have a separation between `core` and `custom` config patterns so it's clear which patterns have been changed by the user:
```
core_patterns = {
            "catalog": ["catalog*", "catalog*/**", "**/catalog*"],
            "parameters": ["parameters*", "parameters*/**", "**/parameters*"],
            "credentials": ["credentials*", "credentials*/**", "**/credentials*"],
            "logging": ["logging*", "logging*/**", "**/logging*"],
        }

if custom_patterns:
    core_patterns.update(custom_patterns)
self.config_patterns = core_patterns
```

But perhaps this way it's also clear that you can actually update the core patterns as well as add your own, e.g. spark config.

File: kedro/config/abstract_config.py (hunk 5 above)
Comment (idanov) [reply kedro/config/abstract_config.py:28]: Agree, let's do it with `UserDict` then.

Comment (MerelTheisenQB) [reply kedro/config/abstract_config.py]: Done!

**Review by noklam — APPROVED** [review]
Nice PR! I think the development notes needed to be update a bit. I would add that the `config_pattern` is now an argument, although it's not configurable in `settings.py` yet. Maybe we should update the `RELEASE.md` too?

File: RELEASE.md [hunk 7]
```diff
@@ -11,6 +11,7 @@
 # Upcoming Release 0.18.4
 
 ## Major features and improvements
+* Made configuration patterns accessible with dictionary get syntax, by making `AbstractConfigLoader` inherit from `UserDict`. The hardcoded patterns are moved from the `KedroContext` to the config loader classes instead.
```
Comment (idanov) [inline RELEASE.md:14]: I would rephrase it to something more user-friendly, e.g. along the lines of:
```suggestion
* The config loader objects now implement `UserDict` and the configuration is accessed through `conf_loader['catalog']`
* You can configure config file patterns through `settings.py` without creating a custom config loader
```

**Review by idanov — APPROVED** [review]
Awesome! Should we add a small section in the docs to show how to change file patterns through `settings.py`?

**Comment by MerelTheisenQB** [comment]
> Awesome! Should we add a small section in the docs to show how to change file patterns through `settings.py`?

I was planning on doing this in the follow up tasks:
https://github.com/kedro-org/kedro/issues/1864
https://github.com/kedro-org/kedro/issues/1462
